### ISSUE TITLE
wrong inference results with Yi-6B-Chat and Yi-VL-6B

### ISSUE THREAD

**Issue by chenpei2** [ggerganov/llama.cpp#6577]
Steps:
1. Use convert-hf-to-gguf.py to convert Yi-6B-Chat files to gguf file.
2. Run with command

> ./build/bin/main -ngl 33 -m /code/Yi-6B-Chat/ggml-model-f32.gguf -p "once upon a time"

It output some messy code and got stuck.

> Log start
main: build = 2629 (87fb5b42)
main: built with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
main: seed  = 1712713739
llama_model_loader: loaded meta data with 22 key-value pairs and 291 tensors from /code/Yi-6B-Chat/ggml-model-f32.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = LLaMA v2
llama_model_loader: - kv   2:                           llama.vocab_size u32              = 64000
llama_model_loader: - kv   3:                       llama.context_length u32              = 4096
llama_model_loader: - kv   4:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   5:                          llama.block_count u32              = 32
llama_model_loader: - kv   6:                  llama.feed_forward_length u32              = 11008
llama_model_loader: - kv   7:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv   8:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   9:              llama.attention.head_count_kv u32              = 4
llama_model_loader: - kv  10:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  11:                       llama.rope.freq_base f32              = 5000000.000000
llama_model_loader: - kv  12:                          general.file_type u32              = 0
llama_model_loader: - kv  13:                       tokenizer.ggml.model str              = llama
llama_model_loader: - kv  14:                      tokenizer.ggml.tokens arr[str,64000]   = ["<unk>", "<|startoftext|>", "<|endof...
llama_model_loader: - kv  15:                      tokenizer.ggml.scores arr[f32,64000]   = [0.000000, 0.000000, 0.000000, 0.0000...
llama_model_loader: - kv  16:                  tokenizer.ggml.token_type arr[i32,64000]   = [2, 3, 3, 3, 3, 3, 1, 1, 1, 3, 3, 3, ...
llama_model_loader: - kv  17:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  18:                tokenizer.ggml.eos_token_id u32              = 2
llama_model_loader: - kv  19:               tokenizer.ggml.add_bos_token bool             = false
llama_model_loader: - kv  20:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  21:                    tokenizer.chat_template str              = {% if not add_generation_prompt is de...
llama_model_loader: - type  f32:  291 tensors
llm_load_vocab: mismatch in special tokens definition ( 498/64000 vs 267/64000 ).
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = SPM
llm_load_print_meta: n_vocab          = 64000
llm_load_print_meta: n_merges         = 0
llm_load_print_meta: n_ctx_train      = 4096
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 4
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 8
llm_load_print_meta: n_embd_k_gqa     = 512
llm_load_print_meta: n_embd_v_gqa     = 512
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 11008
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 5000000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_yarn_orig_ctx  = 4096
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 7B
llm_load_print_meta: model ftype      = all F32
llm_load_print_meta: model params     = 6.06 B
llm_load_print_meta: model size       = 22.58 GiB (32.00 BPW)
llm_load_print_meta: general.name     = LLaMA v2
llm_load_print_meta: BOS token        = 1 '<|startoftext|>'
llm_load_print_meta: EOS token        = 2 '<|endoftext|>'
llm_load_print_meta: UNK token        = 0 '<unk>'
llm_load_print_meta: LF token         = 315 '<0x0A>'
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA A800 80GB PCIe, compute capability 8.0, VMM: yes
llm_load_tensors: ggml ctx size =    0.22 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 33/33 layers to GPU
llm_load_tensors:        CPU buffer size =  1000.00 MiB
llm_load_tensors:      CUDA0 buffer size = 22121.02 MiB
..............................................................................................
llama_new_context_with_model: n_ctx      = 512
llama_new_context_with_model: n_batch    = 512
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:      CUDA0 KV buffer size =    32.00 MiB
llama_new_context_with_model: KV self size  =   32.00 MiB, K (f16):   16.00 MiB, V (f16):   16.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     0.24 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =   133.00 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =     9.01 MiB
llama_new_context_with_model: graph nodes  = 1030
llama_new_context_with_model: graph splits = 2

> system_info: n_threads = 56 / 112 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 1 | AVX512_VNNI = 1 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 |
> sampling:
>         repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
>         top_k = 40, tfs_z = 1.000, top_p = 0.950, min_p = 0.050, typical_p = 1.000, temp = 0.800
>         mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
> sampling order:
> CFG -> Penalties -> top_k -> tfs_z -> typical_p -> top_p -> min_p -> temperature
> generate: n_ctx = 512, n_batch = 2048, n_predict = -1, n_keep = 0
> 
> 
>  once upon a time, there was a man named Bill who lived in a small town. He was known for his kindness and his willingness to help others. Bill had a heart of gold and a kind word for everyone he met.
> 
> One day, a stranger named John walked into the town. John was lost and seemed very confused. He asked for directions to the nearest inn, but no one seemed to understand him. Bill, who was passing by, heard his question and stopped to help.
> 
> Bill approached John and asked him if he needed any assistance. John, still feeling lost and alone, started to explain his situation. He had come from a faraway land and was looking for the inn where he was supposed to stay for the night.
> 
> Bill listened patiently and offered to show John the way. He led him through the streets of the town, pointing out landmarks and helping him to understand the local dialect. After a while, they reached the inn and John thanked Bill profusely for his help.
> 
> As they parted ways, John told Bill that he was truly grateful for his kindness. He said that in a world filled with strangers, it was rare to find someone so willing to lend a hand. Bill smiled and said that it was his pleasure, adding that helping others was what made life worth living.
> 
> From that day on, Bill became even more known for his generosity and his ability to make others feel welcome and understood. His heart of gold was a beacon in the small town, and everyone looked up to him as a role model for kindness.
> 
> And so, the story of Bill, the man with the heart of gold, was passed down through the generations, inspiring others to be kind and to help those in need.<|im_end|>
> 
> ---
> 
> **Lesson:**
> 
> The moral of the story is that kindness and generosity can go a long way, both in making others feel appreciated and in creating a positive impact on one's community. It's a reminder that small acts of kindness can have a profound effect on those around us, and that by being a good person, we can make a difference in the world.<|im_end|>
> 
> ---
> 
> **Application:**
> 
> In our daily lives, we can apply this moral by being more mindful of the way we treat others. We can choose to show kindness to those around us, whether it's a simple smile, a helping hand, or a listening ear. By doing so, we can make someone's day a,,,,,<|im_end|><|im_end|>&#

With Yi-VL-6B, there were some errors when converting with convert.py.

> Traceback (most recent call last):
  File "/code/llama.cpp/./convert.py", line 1548, in <module>
    main()
  File "/code/llama.cpp/./convert.py", line 1534, in main
    model   = convert_model_names(model, params, args.skip_unknown)
  File "/code/llama.cpp/./convert.py", line 1297, in convert_model_names
    raise ValueError(f"Unexpected tensor name: {name}. Use --skip-unknown to ignore it (e.g. LLaVA)")
ValueError: Unexpected tensor name: model.vision_tower.vision_tower.vision_model.embeddings.class_embedding. Use --skip-unknown to ignore it (e.g. LLaVA)

After adding `--skip_unkown`, a gguf file was generated successfully.But it output a lot of space and messy code when using it to inference.

**Comment by FSSRepo** [comment]
Yi-VL-6B is a multimodal model, obviusly it can't be used as a LLM, this one expects an input of image embeddings to generate a response.

**Comment by chenpei2** [comment]
> Yi-VL-6B is a multimodal model, obviusly it can't be used as a LLM, this one expects an input of image embeddings to generate a response.

Well, Yi-6B-Chat is not a multimodal model

**Comment by Jeximo** [comment]
> And there are some message when convert Yi-VL-6B, will it influence the inference results?

Yes, `--skip-unknown` excluded a vision embedding.

> Well, Yi-6B-Chat is not a multimodal model ... It output some messy code and got stuck.

It appeared to work as expected. It stopped because `n_ctx = 512`, and the output is greater than 512 tokens.

https://github.com/ggerganov/llama.cpp/blob/4e9a7f7f7fb6acbddd1462909c8d696e38edbfcc/examples/main/README.md?plain=1#L73

**Comment by Jeximo** [comment]
> it should be truncated and stop rather than outputing messy code and getting stuck

It continued because default `--n-predict`(*default -1*). Default value will enable infinite text generation, `--n-predict -2` will stop generation immediately when the context is filled.

https://github.com/ggerganov/llama.cpp/blob/4e9a7f7f7fb6acbddd1462909c8d696e38edbfcc/examples/main/README.md?plain=1#L165

> There seems to have something to do with "stop words",any ideas about how to set stop words when running with llama.cpp?

It's unrelated to stopping after filled context. For example, `-r "User:"`. See `-r PROMPT, --reverse-prompt PROMPT`: [Specify reverse prompts to pause text generation](https://github.com/ggerganov/llama.cpp/blob/4e9a7f7f7fb6acbddd1462909c8d696e38edbfcc/examples/main/README.md?plain=1#L101).

> Does that mean there is something that should be done to convert Yi-VL models more correctly?

Sorry, it may have converted proplerly, I'm uncertain. The point is that Yi VL is a multimodal model, so try `llava-cli` instead of `main`. Here's an example: [llava cli prompting](https://github.com/ggerganov/llama.cpp/tree/400d5d722d7edf7de0cf24a18c42b183c65047d2/examples/llava#llava-cli-templating-and-llava-16-prompting).

**Comment by chenpei2** [comment]
> It continued because --n-predict(default -1). Default value will enable infinite text generation, --n-predict -2 will stop generation immediately when the context is filled.

I tried `--n-predict` and got a very short output text,that's not what I need. I got some official answer from 0Yi team that the `"<|im_end|>"` output is related to stop words.I added it to stop words in ollama's modelfile and it worked, but I don't know how to set it with llama.cpp.
And I got some unexpected "###" in the output when using Yi-VL-6B so I guess there's something wrong with the converted gguf file.
